src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,174 @@
+            Documents = [documentFileInfo],
+            AdditionalDocuments = [],
+            AnalyzerConfigDocuments = [],
+            ProjectReferences = [],
+            PackageReferences = [],
+            ProjectCapabilities = [],
+            ContentFilePaths = [],
+            FileGlobs = []
+        };
+
+        var projectSet = AddLoadedProjectSet(documentPath);
+        Project workspaceProject;
+        using (await projectSet.Semaphore.DisposableWaitAsync())
+        {
+            var loadedProject = await this.CreateAndTrackInitialProject_NoLockAsync(projectSet, documentPath, language: projectLanguageName);
+            await loadedProject.UpdateWithNewProjectInfoAsync(projectFileInfo, hasAllInformation: false, _logger);
+
+            ProjectsToLoadAndReload.AddWork(new ProjectToLoad(documentPath, ProjectGuid: null, ReportTelemetry: true));
+            loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(new ProjectToLoad(documentPath, ProjectGuid: null, ReportTelemetry: false));
+            workspaceProject = ProjectFactory.Workspace.CurrentSolution.GetRequiredProject(loadedProject.ProjectId);
+        }
+
+        var document = workspaceProject.Documents.Single();
+        Contract.ThrowIfFalse(document.FilePath == documentPath);
+        return document;
+    }
+
+    public async ValueTask TryRemoveMiscellaneousDocumentAsync(DocumentUri uri, bool removeFromMetadataWorkspace)
+    {
+        var documentPath = uri.ParsedUri is { } parsedUri ? ProtocolConversions.GetDocumentFilePathFromUri(parsedUri) : uri.UriString;
+        await TryUnloadProjectSetAsync(documentPath);
+
+        // also do an unload in case this was the non-file scenario
+        if (removeFromMetadataWorkspace && uri.ParsedUri is not null && _metadataAsSourceFileService.TryRemoveDocumentFromWorkspace(ProtocolConversions.GetDocumentFilePathFromUri(uri.ParsedUri)))
+        {
+            return;
+        }
+
+        var matchingDocument = Workspace.CurrentSolution.GetDocumentIds(uri).SingleOrDefault();
+        if (matchingDocument != null)
+        {
+            var project = Workspace.CurrentSolution.GetRequiredProject(matchingDocument.ProjectId);
+            Workspace.OnProjectRemoved(project.Id);
+        }
+    }
+
+    protected override async Task<(RemoteProjectFile? projectFile, bool hasAllInformation, BuildHostProcessKind preferred, BuildHostProcessKind actual)> TryLoadProjectAsync(
+        BuildHostProcessManager buildHostProcessManager, string documentPath, CancellationToken cancellationToken)
+    {
+        const BuildHostProcessKind buildHostKind = BuildHostProcessKind.NetCore;
+        var buildHost = await buildHostProcessManager.GetBuildHostAsync(buildHostKind, cancellationToken);
+        Contract.ThrowIfFalse(Path.GetExtension(documentPath) == ".cs");
+
+        var fakeProjectPath = VirtualProject.GetVirtualProjectPath(documentPath);
```

Comment:
are we using separate types for these paths?  otherwise, are we going to suffer from teh tainted string problem, where all the apis use strings, and it's trivial to be passing the wrong string around?

Reply:
also, can we call this 'virtualProjectPath'.  If we want to call it 'fake', can we rename the helper to GetFakeProjectPath?

Also, can we doc this?

Reply:
GetVirtualProjectPath itself is doc'd, is there more you were looking for as far as docs?

Reply:
1. i find name mismatches confusing.  For example, what makes thsi project path 'fake'?  If 'fake' is the terminology we're using, why isn't it 'GetFakeProjectPath'.
2. When i mean 'docs', i don't mean 'document the method being called'.  i mean 'document in this code why we're going from documentPath to fakeProjectPath'.  Basically, i don't know why this line of code exists at thsi location.  If it were missing, i'd have no sense of why that was wrong, or why this is right.

'2' is why i keep hammering on using different 'types' to represent the different paths.  Then the code won't even compile because it would be saying "i need a VirtualProjectPath, but you have a WorkspaceBlahBlahPath" and it would be clear a change in domains was needed to even make this sensible.

Reply:
I have updated the names of locals and added comments.